File: src/main/java/org/tensorflow/keras/datasets/Iris.java
 package org.tensorflow.keras.datasets;

 import org.tensorflow.Tensors;
 import org.tensorflow.data.GraphLoader;
 import org.tensorflow.data.GraphModeTensorFrame;
 import org.tensorflow.keras.utils.DataUtils;
 import org.tensorflow.keras.utils.Keras;
 import org.tensorflow.utils.Pair;

 import java.io.BufferedReader;
 import java.io.FileReader;
 import java.io.IOException;
 import java.util.Arrays;

 public class Iris {
    private static final String IRIS_ORIGIN =
            "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data";
    private static final int NUM_EXAMPLES = 151;
    private static final int INPUT_LENGTH = 4;
    private static final int OUTPUT_LENGTH = 3;
    private static final String LOCAL_PREFIX = "datasets/iris/";
    private static final String LOCAL_FILE = "iris.data";

    private enum COLOR {
        setosa(0), versicolor(1), virginica(2);
        private final int value;

        COLOR(int value) { this.value = value; }
        int getValue() { return this.value; }
    }


    public static void main(String[] args) throws IOException {

    }


    public static void download() throws IOException {
        DataUtils.getFile(LOCAL_PREFIX + LOCAL_FILE, IRIS_ORIGIN);
    }

    public static Pair<GraphLoader<Float>, GraphLoader<Float>> loadData(double val_split) throws IOException {
        Iris.download();
        try (BufferedReader br = new BufferedReader(new FileReader(
                Keras.kerasPath(LOCAL_PREFIX + LOCAL_FILE).toFile()))) {
            float[][] X = new float[NUM_EXAMPLES][INPUT_LENGTH];
            float[][] y = new float[NUM_EXAMPLES][OUTPUT_LENGTH];

            assert X.length == y.length;
            int trainSize = (int) (NUM_EXAMPLES * (1 - val_split));

            float[][] XTrain = new float[trainSize][INPUT_LENGTH];
            float[][] yTrain = new float[trainSize][OUTPUT_LENGTH];

            float[][] XVal = new float[NUM_EXAMPLES - trainSize][INPUT_LENGTH];
            float[][] yVal = new float[NUM_EXAMPLES - trainSize][OUTPUT_LENGTH];

            String line;
            int count = 0;
            while((line = br.readLine()) != null && count < trainSize) {
                if (line.equals("")) break;

                String[] values = line.split(",Iris-");

                String[] xstring = values[0].split(",");
                float[] xvector = new float[xstring.length];
                for (int i = 0; i < xstring.length; i++) {
                    xvector[i] = Float.parseFloat(xstring[i]);
                }

                float[] yvector = oneHot(COLOR.valueOf(values[1]).getValue(),
 COLOR.values().length);

                XTrain[count] = xvector;
                yTrain[count] = yvector;
                count ++;
            }

            while((line = br.readLine()) != null && count < NUM_EXAMPLES) {
                if (line.equals("")) break;

                String[] values = line.split(",Iris-");

                String[] xstring = values[0].split(",");
                float[] xvector = new float[xstring.length];
                for (int i = 0; i < xstring.length; i++) {
                    xvector[i] = Float.parseFloat(xstring[i]);
                }

                float[] yvector = oneHot(COLOR.valueOf(values[1]).getValue(),
 COLOR.values().length);

                XVal[count - trainSize] = xvector;
                yVal[count - trainSize] = yvector;
                count ++;
            }



            return new Pair<>(
                    new GraphModeTensorFrame<>(Float.class, Tensors.create(XTrain), Tensors.create(yTrain)),
                    new GraphModeTensorFrame<>(Float.class, Tensors.create(XVal), Tensors.create(yVal))
            );
        }
    }

    private static float[] oneHot(int label, int total) {
        if (label >= total) {
            throw new IllegalArgumentException("Invalid Index for One-Hot Vector");
        }

        float[] oneHot = new float[total];
        Arrays.fill(oneHot, 0);
        oneHot[label] = 1.0f;
        return oneHot;
    }
 }

File: src/main/scala/org/tensorflow/keras/datasets/Iris.scala
package org.tensorflow.keras.datasets

import org.tensorflow.data.{GraphLoader, GraphModeTensorFrame}
import org.tensorflow.keras.utils.{DataUtils, Keras}
import org.tensorflow.types.TFloat32
import org.tensorflow.utils.Tensors

import java.io.{BufferedReader, FileReader, IOException}
import java.util

object Iris {
  private val IRIS_ORIGIN   = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
  private val NUM_EXAMPLES  = 151
  private val INPUT_LENGTH  = 4
  private val OUTPUT_LENGTH = 3
  private val LOCAL_PREFIX  = "datasets/iris/"
  private val LOCAL_FILE    = "iris.data"

  private object COLOR extends Enumeration {
    type COLOR = Value
    val setosa, versicolor, virginica = Value

    def valueOf(name: String): COLOR = name match {
      case "setosa"     => setosa
      case "versicolor" => versicolor
      case "virginica"  => virginica
    }

//    private val value = 0d ef this (value: Int) {
//      this ()
//      this.value = value
//    }
//
//    private[datasets] def getValue = this.value
  }

  @throws[IOException]
  def main(args: Array[String]): Unit = ()

  @throws[IOException]
  def download(): Unit =
    DataUtils.getFile(LOCAL_PREFIX + LOCAL_FILE, IRIS_ORIGIN)

  @throws[IOException]
  def loadData(val_split: Double): (GraphLoader[TFloat32], GraphLoader[TFloat32]) = {
    Iris.download()
    val fr = new FileReader(Keras.kerasPath(LOCAL_PREFIX + LOCAL_FILE).toFile)
    try {
      val br = new BufferedReader(fr)
      try {
        val X = Array.ofDim[Float](NUM_EXAMPLES, INPUT_LENGTH)
        val y = Array.ofDim[Float](NUM_EXAMPLES, OUTPUT_LENGTH)
        val trainSize = (NUM_EXAMPLES * (1 - val_split)).toInt
        val XTrain = Array.ofDim[Float](trainSize, INPUT_LENGTH)
        val yTrain = Array.ofDim[Float](trainSize, OUTPUT_LENGTH)
        val XVal = Array.ofDim[Float](NUM_EXAMPLES - trainSize, INPUT_LENGTH)
        val yVal = Array.ofDim[Float](NUM_EXAMPLES - trainSize, OUTPUT_LENGTH)
        var line: String = null
        var count = 0
        while ({
          line = br.readLine()
          line != null && count < trainSize && line != ""
        }) {
          val values    = line.split(",Iris-")
          val xstring   = values(0).split(",")
          val xvector   = new Array[Float](xstring.length)
          for (i <- 0 until xstring.length) {
            xvector(i) = java.lang.Float.parseFloat(xstring(i))
          }
          val yvector   = oneHot(COLOR.valueOf(values(1)).id, COLOR.values.size)
          XTrain(count) = xvector
          yTrain(count) = yvector
          count += 1
        }

        while ({
          line = br.readLine()
          line != null && count < NUM_EXAMPLES && line != ""
        }) {
          val values = line.split(",Iris-")
          val xstring = values(0).split(",")
          val xvector = new Array[Float](xstring.length)
          for (i <- 0 until xstring.length) {
            xvector(i) = java.lang.Float.parseFloat(xstring(i))
          }
          val yvector = oneHot(COLOR.valueOf(values(1)).id, COLOR.values.size)
          XVal(count - trainSize) = xvector
          yVal(count - trainSize) = yvector
          count += 1
        }

        (
          new GraphModeTensorFrame(classOf[TFloat32], Tensors.create(XTrain), Tensors.create(yTrain)),
          new GraphModeTensorFrame(classOf[TFloat32], Tensors.create(XVal)  , Tensors.create(yVal))
        )
      } finally br.close()
    } finally fr.close()
  }

  private def oneHot(label: Int, total: Int) = {
    if (label >= total) throw new IllegalArgumentException("Invalid Index for One-Hot Vector")
    val oneHot = new Array[Float](total)
    util.Arrays.fill(oneHot, 0)
    oneHot(label) = 1.0f
    oneHot
  }
}
